---
title: Community network plots
author: Gibran Hemani
data: 25/02/2019
---

```{r}
library(knitr)
opts_chunk$set(cache=FALSE, echo=TRUE, message=FALSE, warning=FALSE)
```

Plot:

- community 2 = largest
- community 9 = cohesin
- community 22 = nfkb
- community 76 = ezh2

```{r}
library(dplyr)
library(googleVis)
library(networkD3)
library(rbokeh)
library(htmlwidgets)
library(igraph)
library(purrr)
library(ggbio)
library(GenomicRanges)
# Also need to install phantomjs
# brew tap homebrew/cask
# brew cask install phantomjs


make_sankey <- function(userset, out, snpanno, options = list(height=600, width=700), method=c("googlevis", "networkd3")[1])
{
	temp <- subset(out, userSet == userset) %>% mutate(snp=gsub(":SNP", "", snp))
	snpanno <- snpanno %>% ungroup %>% mutate(snp=gsub(":SNP", "", snp))
	out <- mutate(out, snp=gsub(":SNP", "", snp))
	bigdat <- list(
		filter(snpanno, snp %in% temp$snp) %>% 
			ungroup() %>% 
			select(start=antibody, end=snp, weight=n) %>%
			mutate(section=1),
		filter(temp, !duplicated(paste(snp, cpg))) %>%
			ungroup() %>%
			mutate(weight=as.numeric(cis) * -1 + 2, section=2) %>%
			select(start=snp, end=cpg, weight=weight, section=section),
		temp %>% ungroup %>%
			mutate(section=3) %>%
			select(start=cpg, end=antibody, weight=pValueLog, section=section)
	) %>% bind_rows()
	# If any snp has no annotation, give it one
	snplist <- unique(temp$snp)
	missnp <- snplist[!snplist %in% subset(bigdat, section == 1)$end]
	if(length(missnp) > 0)
	{
		message(missnp)
		bigdat <- bind_rows(bigdat, data_frame(start="No annotation", end=missnp, weight=1, section=1)) %>% arrange(section)
	}
	cpglist <- unique(temp$cpg)
	miscpg <- cpglist[!cpglist %in% subset(bigdat, section == 3)$start]
	if(length(miscpg) > 0)
	{
		message(miscpg)
		bigdat <- bind_rows(bigdat, data_frame(start=miscpg, end="No annotation", weight=1, section=3)) %>% arrange(section)
	}

	bigdat$end[bigdat$section==3] <- paste0(bigdat$end[bigdat$section==3], " ")
	if(method == "googlevis")
	{
		a <- gvisSankey(subset(bigdat), from="start", to="end", weight="", options=options)
		a$html$footer <- "</body>\n</html>"
		a$html$caption <- ""
		return(a)
	} else if(method == "networkd3")
	{
		nom <- tibble(name=unique(c(bigdat$start, bigdat$end)))
		nom$index <- 1:nrow(nom)-1

		bigdat <- merge(bigdat, nom, by.x="start", by.y="name")
		names(bigdat)[names(bigdat) == "index"] <- "source"
		bigdat <- merge(bigdat, nom, by.x="end", by.y="name")
		names(bigdat)[names(bigdat) == "index"] <- "target"
		bigdat$value <- 1

		o <- sankeyNetwork(Links=bigdat, Nodes=nom, Source="source", Target="target", Value="value", NodeID="name", fontSize = 18, fontFamily="Arial", height=options$height, width=options$width)
		return(o)
	}
}

```

Load the results from `plot_setup.rmd` and modify the `sheffield_dnase` annotations to be only present for DNAse hypersensitivity in blood

```{r}

load("../results/plotting_community_enrichments.rdata")

snpo2$antibody[is.na(snpo2$antibody)] <- snpo2$description[is.na(snpo2$antibody)]
ind <- snpo2$collection == "sheffield_dnase"
ind2 <- grepl("Hematop", snpo2$antibody) | grepl("sheffield_dnase", snpo2$antibody)
snpo2$antibody[ind] <- NA
snpo2$antibody[ind2] <- "Blood Dnase HS"
snpo2 <- subset(snpo2, !is.na(antibody))
snpanno <- group_by(snpo2, snp, antibody) %>%
	summarise(n=n()) %>%
	arrange(desc(n)) %>%
	filter(between(row_number(), 1, pmin(10, n())))
```

Save plots in `networkD3` format

```{r}
a <- make_sankey(2, out, snpanno, list(height=1600, width=1000), "networkd3")
widget2png(a, "../images/sankey2.png")

a <- make_sankey(22, out, snpanno, list(height=600, width=1000), "networkd3")
widget2png(a, "../images/sankey22.png")

a <- make_sankey(76, out, snpanno, list(height=300, width=1000), "networkd3")
widget2png(a, "../images/sankey76.png")

a <- make_sankey(9, out, snpanno, list(height=700, width=1000), "networkd3")
widget2png(a, "../images/sankey9.png")
```

Save plots in `googleVis` format. This has to be done manually - `plot(a)` will open in a webpage. Print the webpage to pdf and then crop manually. This is how to create the files `../images/gvsankey2.pdf` etc.


```{r}
a <- make_sankey(2, out, snpanno, list(height=1600, width=1000), "googlevis")
plot(a)

a <- make_sankey(22, out, snpanno, list(height=600, width=1000), "googlevis")
plot(a)

a <- make_sankey(76, out, snpanno, list(height=300, width=1000), "googlevis")
plot(a)

a <- make_sankey(9, out, snpanno, list(height=700, width=1000), "googlevis")
plot(a)

```

---

## Network plots

Setup data

```{r}
# load("../results/gwas_clusters_nochr6.rdata")
# info <- read.csv("../../data/gwas/00info.csv")
# info <- data.frame(fn=gsub(".txt.gz", "", info$newfile), id=info$id)

# labels <- subset(res, !duplicated(i), select=c(i, fn))
# labels$fn <- gsub("../../data/gwas/", "", labels$fn)
# labels$fn <- gsub(".txt.gz", "", labels$fn)
# dat <- merge(dat, labels, by.x="id", by.y="i")
# sign<-0.05/nrow(dat)
# g<-grep("metabolites__",dat$fn)
# gwa<-dat[-g,]
# gwa<-gwa[which(gwa$binom4<sign),]
# gwa.clust<-unique(gwa$clust)

rm(list=ls())
load("../results/graph.rdata")

memcount <- group_by(mem, cluster) %>%
	summarise(n=n()) %>%
	arrange(desc(n))

dat <- group_by(dat, creg) %>%
	mutate(snp1=snp[1])

extract_subset <- function(gr, dat, cpg_list)
{
	keep <- subset(dat, creg %in% cpg_list | tcpg %in% cpg_list)
	keep1 <- subset(keep, select=c(creg, tcpg))
	keep1$type <- "CpG"
	keep2 <- data_frame(creg=keep$snp1, tcpg=keep$creg, type="cis")
	keep3 <- data_frame(creg=keep$snp1, tcpg=keep$tcpg, type="trans")
	keep <- bind_rows(keep1, keep2, keep3)
	keep$lab <- paste(keep$creg, keep$tcpg)
	dup <- keep$lab[duplicated(keep$lab)]
	subset(keep, lab %in% dup)
	keep <- subset(keep, !duplicated(lab))
	# keep <- subset(keep, type == "CpG")
	c <- graph_from_data_frame(keep)
	V(c)$what <- "CpG"
	V(c)[grepl("chr", V(c)$name)]$what <- "SNP"
	V(c)$color <- ifelse(grepl("SNP", names(V(c))), 2, 3)
	return(list(gr=c, dat=keep))
}

plot_community <- function(cluster)
{
	temp <- extract_subset(gr, dat, mem$cpg[mem$cluster == cluster])
	plot(temp$gr, 
		layout=layout.fruchterman.reingold, 
		# vertex.color=as.numeric(as.factor(V(temp$gr)$what)), 
		vertex.size=(as.numeric(as.factor(V(temp$gr)$what))-5)^2, 
		vertex.label=NA, 
		edge.arrow.size=0, 
		edge.color=as.numeric(as.factor(E(temp$gr)$type))
	)
}
```

Plot community that contains longest distance

```{r}
(temp <- farthest.nodes(gr))
temp <- all_simple_paths(gr, temp$vertices[1], temp$vertices[2])
temp <- extract_subset(gr, dat, names(temp[[1]]))
pdf(file="../images/plot_longest1.pdf")
plot(temp$gr, 
	layout=layout.fruchterman.reingold, 
	# vertex.color=as.numeric(as.factor(V(temp$gr)$what)), 
	vertex.size=(as.numeric(as.factor(V(temp$gr)$what))-5)^2, 
	vertex.label=NA, 
	edge.arrow.size=0, 
	edge.color=as.numeric(as.factor(E(temp$gr)$type))
)
dev.off()
```

Plot community 6

```{r}
plot_community(6)
```

Plot all communites with >= 5 CpGs

```{r}
todo <- subset(memcount, n >= 5)$cluster
```

This will be `r length(todo)` communities

```{r}
dir.create("../images/community_plots/", recursive=TRUE)
for(i in todo)
{
	message(i)
	temp <- extract_subset(gr, dat, mem$cpg[mem$cluster == i])
	pdf(paste0("../images/community_plots/cluster", i, ".pdf"))
	plot(temp$gr, 
		layout=layout.fruchterman.reingold, 
		# vertex.color=as.numeric(as.factor(V(temp$gr)$what)), 
		vertex.size=(as.numeric(as.factor(V(temp$gr)$what))-5)^2, 
		vertex.label=NA, 
		edge.arrow.size=0, 
		edge.color=as.numeric(as.factor(E(temp$gr)$type))
	)
	dev.off()
}

```


What do these look like on circos plots?

Define chromosome ranges:

```{r}

hg19chr <- GRanges(seqnames=c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr20", "chr19", "chr21", "chr22", "chr23"),
	IRanges(rep(1,23),
		c(249250621,243199373,198022430,191154276,180915260,171115067,155270560,146364022,141213431,135534747,135006516,133851895,115169878,107349540,102531392,90354753,81195210,78077248,63025520,59128983,48129895, 51304566, 159138663)
	)
)
```

This function wrangles the data for a specific community

```{r}
extract_subset_granges <- function(gr, dat, cpg_list, hg19chr)
{
	keep <- subset(dat, creg %in% cpg_list | tcpg %in% cpg_list)
	keep <- tidyr::separate(keep, snp, c("snpchr", "snppos", "snptype"), sep=":", remove=FALSE)
	g1 <- GRanges(seqnames=keep$creg_chr, IRanges(keep$creg_pos, keep$creg_pos))
	g2 <- GRanges(seqnames=keep$tcpg_chr, IRanges(keep$tcpg_pos, keep$tcpg_pos))
	g1 <- c(hg19chr, g1)
	g2 <- c(hg19chr, g2)
	g1$topos <- g2
	g1 <- g1[-c(1:length(hg19chr))]

	temp <- subset(keep, !duplicated(snp))
	g3 <- GRanges(seqnames=temp$snpchr, IRanges(as.numeric(temp$snppos), as.numeric(temp$snppos)))
	g3$what <- "snp"
	temp <- subset(keep, !duplicated(creg))
	g4 <- GRanges(seqnames=temp$creg_chr, IRanges(as.numeric(temp$creg_pos), as.numeric(temp$creg_pos)))
	g4$what <- "creg"
	temp <- subset(keep, !duplicated(tcpg))
	g5 <- GRanges(seqnames=temp$tcpg_chr, IRanges(as.numeric(temp$tcpg_pos), as.numeric(temp$tcpg_pos)))
	g5$what <- "tcpg"

	g3$y <- 0
	g4$y <- 0
	g5$y <- 0
	pos <- c(g3,g4,g5)
	h <- hg19chr
	h$what <- NA
	h$y <- NA
	pos <- c(h, pos)
	return(list(links=g1, pos=pos))
}

plot_circle <- function(temp, hg19chr)
{
	ggplot() + 
		layout_circle(temp$pos[1:23], geom = "ideo", fill = "black", radius = 8.4, trackWidth = 0.2) +
		layout_circle(temp$links, geom = "link", linked.to = "topos", radius=6) +
		layout_circle(temp$pos[-c(1:23)], geom = "point", aes(x=y, colour=what), size=6, radius=6) + 
		scale_colour_manual(values=c("#9ecae1", "#a1d99b", "#31a354")) +
		theme(legend.position="none")

}

```

Do all communities with 5 or more CpGs


```{r}
todo <- subset(memcount, n >= 5)$cluster
for(i in todo)
{
	message(i)
	temp <- extract_subset_granges(gr, dat, mem$cpg[mem$cluster == i], hg19chr)
	pdf(paste0("../images/community_plots/cluster_circle", i, ".pdf"))
	p <- plot_circle(temp, hg19chr)
	print(p)
	dev.off()
}
```

## Full graph plots

Get hub and authority scores for each CpG

```{r}
hub <- hub.score(gr)$vector
auth <- authority.score(gr)$vector

plot(log(hub), log(auth))
```

Try plotting all

```{r}
plot(gr, layout=layout.drl, vertex.color=abs(as.numeric(hub > 1)-2), vertex.frame.color=abs(as.numeric(hub > 1)-2), vertex.size = hub, vertex.label = NA, edge.arrow.size = 0, edge.width=1, edge.color="black")
```

Looks busy.. Try extracting only communities with 10 or more CpGs


```{r}

com <- dplyr::as_data_frame(mem)
names(com) <- c("cpg", "mem")
a <- group_by(com, mem) %>%
	summarise(n=n()) %>%
	filter(n >= 10)

comkeep <- subset(com, mem %in% a$mem)
remcpg <- com$cpg[!com$cpg %in% comkeep$cpg]
grc <- delete_vertices(gr, remcpg)

hub <- (hub.score(grc)$vector + 2)^2 - 1
ind <- match(comkeep$cpg, V(grc)$name)
all(V(grc)$name == comkeep$cpg)
V(grc)$community <- comkeep$mem
V(grc)$community2 <- V(grc)$community
V(grc)$community2[duplicated(V(grc)$community2)] <- NA
pdf(file="../images/community_plot_10plus.pdf")
plot(grc, layout=layout.fruchterman.reingold, vertex.color=V(grc)$community, vertex.frame.color=V(grc)$community, vertex.size = 2, vertex.label = V(grc)$community2, edge.arrow.size = 0)
dev.off()

```

Try with only 15 or more

```{r}
com <- dplyr::as_data_frame(mem)
names(com) <- c("cpg", "mem")
a <- group_by(com, mem) %>%
	summarise(n=n()) %>%
	filter(n >= 15)

comkeep <- subset(com, mem %in% a$mem)
remcpg <- com$cpg[!com$cpg %in% comkeep$cpg]
grc <- delete_vertices(gr, remcpg)

hub <- (hub.score(grc)$vector + 2)^2 - 1
ind <- match(comkeep$cpg, V(grc)$name)
all(V(grc)$name == comkeep$cpg)
V(grc)$community <- comkeep$mem
V(grc)$community2 <- V(grc)$community
V(grc)$community2[duplicated(V(grc)$community2)] <- NA
pdf(file="../images/community_plot_15plus.pdf")
plot(grc, layout=layout.fruchterman.reingold, vertex.color=V(grc)$community, vertex.frame.color=V(grc)$community, vertex.size = 2, vertex.label = V(grc)$community2, edge.arrow.size = 0)
dev.off()
```

20 or more...

```{r}
com <- dplyr::as_data_frame(mem)
names(com) <- c("cpg", "mem")
a <- group_by(com, mem) %>%
	summarise(n=n()) %>%
	filter(n >= 20)

comkeep <- subset(com, mem %in% a$mem)
remcpg <- com$cpg[!com$cpg %in% comkeep$cpg]
grc <- delete_vertices(gr, remcpg)

hub <- (hub.score(grc)$vector + 2)^2 - 1
ind <- match(comkeep$cpg, V(grc)$name)
all(V(grc)$name == comkeep$cpg)
V(grc)$community <- comkeep$mem
V(grc)$community2 <- V(grc)$community
V(grc)$community2[duplicated(V(grc)$community2)] <- NA
pdf(file="../images/community_plot_20plus.pdf")
plot(grc, layout=layout.fruchterman.reingold, vertex.color=V(grc)$community, vertex.frame.color=V(grc)$community, vertex.size = (as.numeric(!is.na(V(grc)$community2))) * 5, vertex.label = V(grc)$community2, edge.arrow.size = 0, vertex.label.family="sans", vertex.label.color="black", vertex.label.cex=0.5)
dev.off()

```
