---
title: Trans-mQTL analysis of CPG to trait follow up
author: Gibran Hemani
date: "`r format(Sys.time(), '%d %B, %Y')`"
---

```{r}
library(knitr)
library(dplyr)
library(TwoSampleMR)
library(ggplot2)
library(ggthemes)

threshold1 <- 0.05 / (300000 * 698)
threshold2 <- 0.05 / (300000)

fn <- read.csv("../../data/gwas/00info.csv")
load("../results/to_extractt.rdata")
load("../results/to_extractc.rdata")
load("../results/cpg_trait_coloc.rdata")
load("../../data/misc/cpg_pos.rdata")
load("../results/trans_extract.rdata")
load("../results/cis_extract.rdata")
load("../../results/16/16_clumped.rdata")
load("../data/snp_1kg.rdata")
snp_1kg <- with(snp_1kg, data_frame(snpchr=V1, snppos=V4, snp2=snp, snp=V2))

ao <- available_outcomes() %>% filter(id %in% res$outcome)
sig <- res[res$H4 > 0.8 & res$p < threshold1,]
sig <- subset(sig, !duplicated(paste(exposure, outcome)))
sig <- merge(sig, subset(ao, select=c(id, trait, category, subcategory)), by.x="outcome", by.y="id")
sig <- inner_join(sig, cpgpos, by=c("exposure"="cpg"))
sig$cpgchrn <- as.numeric(gsub("chr", "", sig$cpgchr))

sig <- inner_join(sig, snp_1kg, by=c("snp"))
dim(sig)

table(sig$snpchr == sig$cpgchrn, sig$nsnp < 10)

```

We have a number of putative CpG-trait associations based on colocalisation at the of the cis mQTLs. Some of these CpGs also have trans effects. We can use the trans effects as replication of the results.

```{r}

temp1 <- select(to_extractt, transsnp, outcome, snp, exposure, b, se, p, n)
trans_extract$code <- paste(trans_extract$id.outcome, trans_extract$transsnp)
temp1$code <- paste(temp1$outcome, temp1$transsnp)
table(temp1$code %in% trans_extract$code)


exposure <- format_data(
	a <- subset(clumped, snp %in% trans_extract$transsnp),
	snp_col="snp",
	phenotype_col="cpg",
	effect_allele_col = "Allele1",
	other_allele_col = "Allele2",
	eaf_col = "Freq1",
	beta_col = "Effect",
	se_col = "StdErr",
	pval_col = "pval",
	samplesize_col = "TotalSampleSize"
)

trans_extract$rsid <- trans_extract$SNP
trans_extract$SNP <- tolower(trans_extract$transsnp)
dat <- harmonise_data(exposure, trans_extract)
dat$code <- paste(dat$exposure, dat$id.outcome)

to_extractt$code <- paste(to_extractt$exposure, to_extractt$outcome)

dat2 <- subset(dat, code %in% to_extractt$code)

table(table(dat2$code))

mr_trans <- mr(dat2, method_list=c("mr_wald_ratio", "mr_ivw"))
mr_trans2 <- subset(mr_trans, method %in% c("Wald ratio", "Inverse variance weighted"))


temp1 <- select(to_extractc, transsnp, outcome, snp, exposure, b, se, p, n)
cis_extract$code <- paste(cis_extract$id.outcome, cis_extract$transsnp)
temp1$code <- paste(temp1$outcome, temp1$transsnp)
table(temp1$code %in% cis_extract$code)


exposure <- format_data(
	a <- subset(clumped, snp %in% cis_extract$transsnp),
	snp_col="snp",
	phenotype_col="cpg",
	effect_allele_col = "Allele1",
	other_allele_col = "Allele2",
	eaf_col = "Freq1",
	beta_col = "Effect",
	se_col = "StdErr",
	pval_col = "pval",
	samplesize_col = "TotalSampleSize"
)

cis_extract$rsid <- cis_extract$SNP
cis_extract$SNP <- tolower(cis_extract$transsnp)
dat_cis <- harmonise_data(exposure, cis_extract)
dat_cis$code <- paste(dat_cis$exposure, dat_cis$id.outcome)

to_extractc$code <- paste(to_extractc$exposure, to_extractc$outcome)

dat_cis2 <- subset(dat_cis, code %in% to_extractc$code)

table(table(dat_cis2$code))

mr_cis <- mr(dat_cis2, method_list=c("mr_wald_ratio", "mr_ivw"))
mr_cis2 <- subset(mr_cis, method %in% c("Wald ratio", "Inverse variance weighted"))


mr_cis_trans <- inner_join(mr_cis, mr_trans, by=c("exposure", "id.outcome"), suffix=c("cis", "trans"))

plot(btrans ~ bcis, mr_cis_trans)

mr_cis_trans$zcis <- mr_cis_trans$bcis / mr_cis_trans$secis
mr_cis_trans$ztrans <- mr_cis_trans$btrans / mr_cis_trans$setrans


pdf("../images/cis-trans.pdf")
plot(ztrans ~ zcis, mr_cis_trans, xlab="Cis effect", ylab="Trans effect")
dev.off()

table(sign(mr_cis_trans$bcis) == sign(mr_cis_trans$btrans))


temp1 <- subset(sig, select=c(outcome, exposure, b, se))
temp2 <- subset(mr3, select=c(id.outcome, exposure, b, se))
temp3 <- inner_join(temp1, temp2, by=c("outcome"="id.outcome", "exposure"="exposure"))

plot(temp3$b.x/temp3$se.x, temp3$b.y/temp3$se.y)

temp2 <- inner_join(temp1, trans_extract, by="code")






load("../results/trans_extract.rdata")
load("../results/cis_extract.rdata")






```

