---
title: MR of CPG to trait analysis
author: Gibran Hemani
date: "`r format(Sys.time(), '%d %B, %Y')`"
---

```{r}
suppressWarnings(suppressPackageStartupMessages({
	library(knitr)
	library(TwoSampleMR)
	library(dplyr)
	library(ggplot2)
	library(ggthemes)
}))
opts_chunk$set(cache=TRUE, echo=TRUE, message=FALSE, warning=FALSE)

ao <- available_outcomes()
threshold1 <- 0.05 / (190000 * 122)
threshold2 <- 0.05 / (190000)

fn <- read.csv("../../data/gwas/00info.csv")

load("../results/cpg_trait_coloc.rdata")
load("../../data/misc/cpg_pos.rdata")
res <- as.data.frame(res, stringsAsFactors=FALSE)
ao <- ao %>% filter(id %in% res$outcome)
sig <- res[res$H4 > 0.8 & res$p < threshold1 & res$nsnp >= 10,]
sig <- subset(sig, !duplicated(paste(exposure, outcome)))
sig <- merge(sig, subset(ao, select=c(id, trait, category, subcategory)), by.x="outcome", by.y="id")
sig <- inner_join(sig, cpgpos, by=c("exposure"="cpg"))
sig$cpgchrn <- as.numeric(gsub("chr", "", sig$cpgchr))
sigm <- subset(sig, category != "Metabolites")

```


How many traits with more than one instrument

```{r}
load("../../06_mr-gwas-cpg/data/snps_gwas.rdata")
am <- filter(a, grepl("mr", data_source.exposure))
gkeep <- am %>% group_by(id.exposure, trait) %>% summarise(n=n()) %>% arrange(desc(n)) %>% ungroup %>% filter(!duplicated(trait))
table(gkeep$n == 1)

```

We can define `CPG->trait` associations to be significant if the colocalisation score `> 0.8` and the p-value adjusted for 300000 CpGs tested in 698 traits (p < `r threshold1`). How many significant associations?

```{r}
nrow(sig)
```

How many unique CpGs?

```{r}
length(unique(sig$exposure))
```

How many unique traits?

```{r}
length(unique(sig$outcome))
```

How many traits can a single CpG influence?

```{r}
hist(table(sig$exposure), breaks=max(table(sig$exposure)))
hist(table(sigm$exposure), breaks=max(table(sigm$exposure)))
```

How many associations per category of trait?

```{r}
kable(as.data.frame(table(sig$category)))
```

Which traits have the most associations?

```{r}
sig %>% filter(category != "Metabolites") %>% group_by(trait) %>% summarise(cat=first(subcategory), n=n()) %>% arrange(desc(n)) %>% kable
```

How many associations per chromosome?

```{r}
sig %>% group_by(cpgchrn) %>% summarise(n=n()) %>% 
ggplot(aes(x=cpgchrn, y=n)) +
geom_point() +
geom_smooth(method="lm")

```

Chromosome 6 clearly has far more associations than expected by chance. Potentially unreliable because colocalisation analysis will not be able perform correctly with high LD regions. 

Which traits are these MHC associations (where the CpG is in the MHC)?

```{r}
load("../../results/16/16_clumped.rdata")
load("../data/snp_1kg.rdata")
table(sig$snp %in% snp_1kg$V2)
temp <- subset(snp_1kg, V2 %in% sig$snp, select=c(V1, V2, V4, snp))
names(temp) <- c("snpchr", "snp", "snppos", "snpid")
sig <- merge(sig, temp, by="snp")

sig %>% 
filter(category != "Metabolites") %>% 
filter(snpchr == 6 & snppos > 25000000 & snppos < 35000000) %>%
group_by(trait) %>% 
summarise(cat=first(subcategory), n=n()) %>% 
arrange(desc(n)) %>% kable

sig6 <- filter(sig, category != "Metabolites" & !(snpchr == 6 & snppos > 28000000 & snppos < 33000000))
sig6$cis <- sig6$cpgchrn == sig6$snpchr & abs(sig6$snppos - sig6$cpgpos) < 1e6

library(magrittr)
sig6csv <- sig6 %$% data_frame(
	cpg=exposure,
	trait=trait,
	cpg_pos = paste0(cpgchr, ":", cpgpos),
	lead_snp = snp,
	cis=cis,
	b_wald_ratio = b,
	se_wald_ratio = se,
	p_wald_ratio = p,
	coloc_nsnp = nsnp,
	coloc_h0 = H0,
	coloc_h1 = H1,
	coloc_h2 = H2,
	coloc_h3 = H3,
	coloc_h4 = H4
)
write.csv(sig6csv, "../results/cpg-trait-tophits.csv")

```


How many of the colocalised SNPs also appear in CpG networks

```{r}
load("../../05_cis-trans-networks/results/graph.rdata")
coloc_snps <- unique(sig6$snpid)
tab <- matrix(c(
	sum(coloc_snps %in% dat$snp), sum(coloc_snps %in% clumped$snp), 
	length(coloc_snps) - sum(coloc_snps %in% dat$snp), length(unique(clumped$snp)) - sum(coloc_snps %in% clumped$snp)
), 2, 2)
fisher.test(tab)
chisq.test(tab)

coloc_cpgs <- unique(sig6$exposure)
comcpg <- c(dat$creg, dat$tcpg)
tabc <- matrix(c(
	sum(coloc_cpgs %in% comcpg), sum(coloc_cpgs %in% clumped$cpg), 
	length(coloc_cpgs) - sum(coloc_cpgs %in% comcpg), length(unique(clumped$cpg)) - sum(coloc_cpgs %in% clumped$cpg)
), 2, 2)
fisher.test(tabc)
chisq.test(tabc)

```


