---
title: Following up colocalisation signals with secondary mQTLs
author: Gibran Hemani
date: "`r format(Sys.time(), '%d %B, %Y')`"
---

```{r}
suppressWarnings(suppressPackageStartupMessages({
	library(knitr)
	library(dplyr)
	library(TwoSampleMR)
	library(ggplot2)
	library(ggthemes)
	library(magrittr)
}))
opts_chunk$set(cache=TRUE, echo=TRUE, message=FALSE, warning=FALSE)
```

```{r}
threshold1 <- 0.05 / (300000 * 698)
threshold2 <- 0.05 / (300000)

fn <- read.csv("../../data/gwas/00info.csv")
load("../results/to_extractt.rdata")
load("../results/to_extractc.rdata")
load("../results/cpg_trait_coloc.rdata")
load("../../data/misc/cpg_pos.rdata")
load("../results/trans_extract.rdata")
load("../results/cis_extract.rdata")
load("../../results/16/16_clumped.rdata")
load("../data/snp_1kg.rdata")
load("../../05_cis-trans-networks/data/labelids.rdata")
res <- as.data.frame(res, stringsAsFactors=FALSE)
# snp_1kg <- with(snp_1kg, data_frame(snpchr=V1, snppos=V4, snp2=snp, snp=V2))

ao <- available_outcomes() %>% filter(id %in% res$outcome, id %in% labelids$id, access != "developer")
sig <- res[res$H4 > 0.8 & res$p < threshold1,]
sig <- subset(sig, !duplicated(paste(exposure, outcome)))
sig <- merge(sig, subset(ao, select=c(id, trait, category, subcategory)), by.x="outcome", by.y="id")
sig <- inner_join(sig, cpgpos, by=c("exposure"="cpg"))
sig$cpgchrn <- as.numeric(gsub("chr", "", sig$cpgchr))

sig <- inner_join(sig, snp_1kg, by=c("snp"))
dim(sig)
sum(sig$H4 > 0.8 & sig$p < threshold1 & sig$nsnp >= 10)

table(sig$snpchr == sig$cpgchrn, sig$nsnp < 10)

```

We have a number of putative CpG-trait associations based on colocalisation. Some of these CpGs also have secondary effects. We can use the secondary effects to replicate the result

```{r}

temp1 <- select(to_extractt, transsnp, outcome, snp, exposure, b, se, p, n)
trans_extract$code <- paste(trans_extract$id.outcome, trans_extract$transsnp)
temp1$code <- paste(temp1$outcome, temp1$transsnp)
table(temp1$code %in% trans_extract$code)

exposure <- format_data(
	a <- subset(clumped, snp %in% trans_extract$transsnp),
	snp_col="snp",
	phenotype_col="cpg",
	effect_allele_col = "Allele1",
	other_allele_col = "Allele2",
	eaf_col = "Freq1",
	beta_col = "Effect",
	se_col = "StdErr",
	pval_col = "pval",
	samplesize_col = "TotalSampleSize"
)

trans_extract$rsid <- trans_extract$SNP
trans_extract$SNP <- tolower(trans_extract$transsnp)
dat <- harmonise_data(exposure, trans_extract %>% filter(id.outcome %in% ao$id))
dat$code <- paste(dat$exposure, dat$id.outcome)

to_extractt$code <- paste(to_extractt$exposure, to_extractt$outcome)

exposure <- format_data(
	a <- subset(clumped, snp %in% cis_extract$transsnp),
	snp_col="snp",
	phenotype_col="cpg",
	effect_allele_col = "Allele1",
	other_allele_col = "Allele2",
	eaf_col = "Freq1",
	beta_col = "Effect",
	se_col = "StdErr",
	pval_col = "pval",
	samplesize_col = "TotalSampleSize"
)

cis_extract$rsid <- cis_extract$SNP
cis_extract$SNP <- tolower(cis_extract$transsnp)
dat_cis <- harmonise_data(exposure, cis_extract %>% filter(id.outcome %in% ao$id))
dat_cis$code <- paste(dat_cis$exposure, dat_cis$id.outcome)

to_extractc$code <- paste(to_extractc$exposure, to_extractc$outcome)
res$code2 <- paste(res$snp, res$exposure, res$outcome)
dat_cis$code2 <- paste(dat_cis$rsid, dat_cis$code)
table(res$code2 %in% dat_cis$code2)

discovery <- bind_rows(
	subset(dat_cis, paste(dat_cis$rsid, dat_cis$code) %in% res$code2) %>% mutate(cistrans="cis"),
	subset(dat, paste(dat$rsid, dat$code) %in% res$code2) %>% mutate(cistrans="trans")
)

mr_discovery <- mr(discovery, method_list=c("mr_wald_ratio", "mr_ivw"))

replication <- bind_rows(
	subset(dat_cis, ! paste(dat_cis$rsid, dat_cis$code) %in% res$code2) %>% mutate(cistrans="cis"),
	subset(dat, ! paste(dat$rsid, dat$code) %in% res$code2) %>% mutate(cistrans="trans")
)

mr_replication <- mr(replication, method_list=c("mr_wald_ratio", "mr_ivw"))
```

Combine discovery and replication

```{r}
mr_disc_repl <- inner_join(mr_discovery, mr_replication, by=c("exposure", "id.outcome"), suffix=c(".disc", ".repl"))
mr_disc_repl$z.disc <- mr_disc_repl$b.disc / mr_disc_repl$se.disc
mr_disc_repl$z.repl <- mr_disc_repl$b.repl / mr_disc_repl$se.repl
mr_disc_repl$code <- paste(mr_disc_repl$id.outcome, mr_disc_repl$exposure)
mr_disc_repl <- subset(mr_disc_repl, abs(mr_disc_repl$z.disc) > qnorm(threshold1, low=FALSE))
mr_disc_repl <- subset(mr_disc_repl, !duplicated(code))
i <- sign(mr_disc_repl$z.disc) == -1
mr_disc_repl$z.disc2 <- abs(mr_disc_repl$z.disc)
mr_disc_repl$z.repl2 <- mr_disc_repl$z.repl
mr_disc_repl$z.repl2[i] <- mr_disc_repl$z.repl2[i] * -1
mr_disc_repl$fdr <- p.adjust(mr_disc_repl$pval.repl, "fdr")
save(mr_discovery, mr_replication, mr_disc_repl, file="../results/mr_disc_repl.rdata")
```

There are `r nrow(mr_disc_repl)` initial colocalisation findings that have secondary mQTLs available (`r {nrow(mr_disc_repl) / nrow(sig) * 100} %>% round(., 1)`%). 


```{r}
ggplot(mr_disc_repl %>% filter(!is.na(z.repl)), aes(y=z.repl, x=z.disc)) +
geom_point(aes(colour=fdr < 0.05)) +
geom_abline(slope=1, linetype="dotted") +
geom_smooth(method="lm") +
ylim(min(mr_disc_repl$z.disc), max(mr_disc_repl$z.disc)) +
labs(x="Z-score primary MR", y="Z-score secondary MR", colour="Replication\nFDR < 0.05")

summary(lm(z.repl ~ z.disc,  mr_disc_repl))
```

Are there more small p-values than expected?

```{r}
binom.test(x=sum(mr_disc_repl$pval.repl < 0.05, na.rm=T), n=nrow(mr_disc_repl), p=0.05)
```

What about direction of effect?

```{r}
binom.test(x=sum(sign(mr_disc_repl$b.disc) == sign(mr_disc_repl$b.repl)), n=nrow(mr_disc_repl), p=0.5)
```

What about direction of effect for just significant assocs?

```{r}
subset(mr_disc_repl, pval.repl < 0.05) %$% binom.test(x=sum(sign(b.disc) == sign(b.repl)), n=length(pval.repl), p=0.5)
```

Is small p-values in repl related to trait type

```{r}
inner_join(mr_disc_repl, subset(ao, select=c(id, subcategory)), by=c("id.outcome"="id")) %$% summary(lm(z.repl ~ subcategory))
```

```{r}
inner_join(mr_disc_repl, subset(ao, select=c(id, subcategory)), by=c("id.outcome"="id")) %>% mutate(sig=pval.repl < 0.05) %$% summary(glm(sig ~ subcategory))
```

```{r}
inner_join(mr_disc_repl, subset(ao, select=c(id, subcategory)), by=c("id.outcome"="id")) %>% mutate(sig=pval.repl < 0.05, blood = subcategory %in% c("Haemotological", "Metal", "Protein")) %$% summary(glm(sig ~ blood))
```

Which associations appear to be working?

```{r}

mr_disc_repl$fdr <- p.adjust(mr_disc_repl$pval.repl, "fdr")
mr_disc_repl$concordant <- sign(mr_disc_repl$b.repl) == sign(mr_disc_repl$b.disc)
subset(mr_disc_repl, fdr < 0.05 & concordant)
subset(mr_disc_repl, fdr < 0.05 & !concordant)


```

Schizophrenia has some consistent results but looking at all of them together is less convincing

```{r}
ggplot(mr_disc_repl %>% filter(id.outcome == 22), aes(y=z.repl, x=z.disc)) +
geom_point(aes(colour=pval.repl < 0.05)) +
geom_abline(slope=1, linetype="dotted") +
geom_smooth(method="lm") +
ylim(min(mr_disc_repl$z.disc), max(mr_disc_repl$z.disc)) +
labs(x="Z-score primary MR", y="Z-score secondary MR", colour="Replication\np-value < 0.05")

```


Create table

```{r}
out <- mr_disc_repl %$% data_frame(
	cpg = exposure,
	outcome = outcome.disc,
	beta.primary = b.disc,
	se.primary = se.disc,
	pval.primary = pval.disc,
	beta.secondary = b.repl,
	se.secondary = se.repl,
	pval.secondary = pval.repl
)

write.csv(out, file="../results/mr_primary_secondary.csv")

```
	

