---
title: Replication of GoDMC results in Generation Scotland
output:
  html_document:
    toc: true
    theme: united
---
We sent Generation Scotland ~285k mQTLs from GoDMC to replicate e.g. these are not filtered on cis pval<1e-8 and trans pval<1e-14 yet.
After filtering on pval we kept 271724 in GoDMC.

Generation Scotland only analysed:

* SNPs (not INDELS)
* chr 1-22
* CpGs that were successfully analysed (i.e. excluding those for which our models had failed to converge etc.)

After filtering on pval and autosomes we kept 252292 in GoDMC.

Generation Scotland have a sample size of ~5000. They had precomputed mQTL results:

* Created methylation residuals accounting for covariates and complex family structure
* Rank transformed residuals CpGs (though may have started with M values)
* Retained SNP-CpG pairs with p < 1e-3

Generation Scotland sent us back 173,645 SNP:CpG pairs with a p-value of <=1e-3.

That is, a SNP that was tested may have been excluded for the following reasons:

* p-value >1e-3
* p-value marked as 'NA' by REGSCAN v0.5. This can occur for the following reasons (given my rudimentary understanding of c++):
* Slope (beta1) estimate of 0.0.
* Minor allele count <=0.0 (of those with a valid phenotype measurement).
* R2 statistic estimate of <=0.0 or >=1.0.
* T estimate of <=0.

Of the 35,389 (209,034-173,645) SNP:CpG pairs not present in our results, 23,917 were tested as the SNP was present in the input genotypes. *NB, this is based on a match of Chromosome and Position only (GRCh 37)*. Please note that there are some SNPs for which we do not have matching alleles. Again, a quick look on Ensembl suggests that these are tri / quad allelic SNPs: you will likely want to exclude these.

This reduced to 209034+(23917-35389)=197,562 SNP-CpG pairs.
Need to remove SNPs with non matching alleles.

*169k were present in both GoDMC and GS*

- 155k cis
- 14k trans


##OLD -see below for update
##n_tested<-209034+(23917-35389)
###173645+23917=197562

---


## Counting and harmonising between GoDMC and GS

```{r}
suppressWarnings(suppressPackageStartupMessages(library(dplyr)))
suppressWarnings(suppressPackageStartupMessages(library(magrittr)))
suppressWarnings(suppressPackageStartupMessages(library(data.table)))
suppressWarnings(suppressPackageStartupMessages(library(ggplot2)))
suppressWarnings(suppressPackageStartupMessages(library(tidyr)))
suppressWarnings(suppressPackageStartupMessages(library(knitr)))
opts_chunk$set(cache=TRUE, echo=FALSE, message=FALSE, warning=FALSE)
```

Read in the replication data from GS

```{r}
gs <- fread("gunzip -c GS_replication.tsv.gz", he=T)
nrow(gs)
```


Filtered to have MAF > 0.01 and imputation quality score > 0.8 & probes that pass det p and nbeads p

```{r}
p <- read.csv("sites_to_exclude_5200_samples.csv")
gs1 <- subset(gs,info>0.8 & freq1>0.01 & freq1<0.99)
```

`r nrow(gs1)` GS assocs remain after SNP QC filtering. Now do probe filtering

```{r}
w <- which(gs1$cpg%in%p$x)
gs1 <- gs1[-w,] #172147
```

`r nrow(gs) - nrow(gs1)` probes failed QC. Final number of GS assocs before merging

```{r}
nrow(gs1)
str(gs1)
```

Assocs in GoDMC that were not tested in GS

```{r}
no_gs <- read.table("godmc_snps_tested_but_not_in_output.tsv.gz",he=T)
str(no_gs)
```

Info about these columns:

* `snpchr`: SNP chromosome, as in the data you provided to us.
* `snppos`: SNP position, as in the data you provided to us.
* `snp`: SNP identifier, as in the data you provided to us.
* `cpg`: CpG identifier, as in the data you provided to us.
* `Allele1`: Allele1, as in the data you provided to us.
* `Allele2`: Allele2, as in the data you provided to us.
* `Freq1`: Freq1, as in the data you provided to us.
* `FreqSE`: FreqSE, as in the data you provided to us.
* `rsid`: rsID of the SNP, as labelled in our data.
* `alleleA`: alleleA, as labelled in our data.
* `alleleB`: alleleB, as labelled in our data.
* `alleleA_frequency`: Frequency of alleleA, in our data, in all 5,101 individuals for which we have methylation data. NB, not all individuals had a methylation value for all CpGs.
* `alleleB_frequency`: Frequency of alleleB, in our data, in all 5,101 individuals for which we have methylation data. NB, not all individuals had a methylation value for all CpGs.


Do filtering on the missed SNPs and probes

```{r}
no_gs <- subset(no_gs, alleleA_frequency>0.01 & alleleA_frequency<0.99 & alleleB_frequency>0.01 & alleleB_frequency<0.99 )
w <- which(no_gs$cpg%in%p$x)
no_gs <- no_gs[-w,]
```

Number of assocs missed in total: `r nrow(no_gs)`

Read in GoDMC assocs

```{r}
load("../results/16/16_clumped.rdata")
clumped<- subset(clumped, pval < 1e-8 & cis | pval < 1e-14 & !cis)

clumped$cist <- "cis"
clumped$cist[!clumped$cis] <- "trans"
clumped <- subset(clumped, snpchr %in% paste0("chr",1:22) & snptype == "SNP")
clumped$code <- clumped %$% paste(snp, cpg)

replication <- inner_join(
	subset(clumped, select=c(snp, cpg, Allele1, Allele2, Freq1, Effect, StdErr, StdErrARE, StdErrMRE, pval, TotalSampleSize, HetDf, HetISq, cis, code)),
	subset(gs1, select=c(snpid, cpg, a1, a0, freq1, beta1, se, p, r2, n, info)),
	by=c("snp"="snpid", "cpg"="cpg")
)
```

After filtering on cis and trans thresholds, indels, autosomes, number of discovery assocs: `r nrow(replication)`.

Some SNPs are mismatching alleles

```{r}
replication$Allele1 <- toupper(replication$Allele1)
replication$Allele2 <- toupper(replication$Allele2)
table(replication$Allele1 == replication$a1 | replication$Allele2 == replication$a1)
```

Remove mismatching SNPs from replication list

```{r}
w <- which(replication$Allele1 == replication$a1 | replication$Allele2 == replication$a1)
mm <- replication[-w,]
replication <- replication[w,]
```

Now harmonise so that the effect is always relative to the same allele in both datasets

```{r}
index <- replication$Allele1 != replication$a1
table(index)
replication$beta1[index] <- replication$beta1[index] * -1
temp <- replication$a0[index]
replication$a0[index] <- replication$a1[index]
replication$a1[index] <- temp
replication$rsqdisc <- 2 * replication$Freq1 * (1-replication$Freq1) * replication$Effect^2
```

Quickly check that the effects align as expected

```{r}
plot(Effect ~ beta1, replication)
```

Final list of replication assocs: `r nrow(replication)`

Which assocs did we provide that was tested and wasn't significant at 1e-3:

```{r}
rep_fail <- inner_join(
	subset(clumped, select=c(snp, cpg, Allele1, Allele2, Freq1, Effect, StdErr, StdErrARE, StdErrMRE, pval, TotalSampleSize, HetDf, HetISq, cis, code)),
	subset(no_gs, select=c(snp, cpg, alleleA, alleleB, alleleB_frequency)),
	by=c("snp"="snp", "cpg"="cpg")
)
nrow(rep_fail)
```

Check alleles

```{r}
rep_fail$Allele1 <- toupper(rep_fail$Allele1)
rep_fail$Allele2 <- toupper(rep_fail$Allele2)
index <- (rep_fail$Allele1 == rep_fail$alleleA & rep_fail$Allele2 == rep_fail$alleleB) | (rep_fail$Allele2 == rep_fail$alleleA & rep_fail$Allele1 == rep_fail$alleleB)
table(index)
rep_fail <- rep_fail[index,]
```

Final list of assocs that were tested but didn't replicate at 1e-3: `r nrow(rep_fail)`

Final list of assocs that attempted replication in GS:

```{r}
n_tested <- nrow(rep_fail)+nrow(replication) 
n_tested
```


## Expected significance

Given the discovery effect sizes, what do we expect the replication rate to be at different replication thresholds (1e-3 and Bonferroni for `r n_tested`)?

What was the replication probability for the ones that failed replication
Total tested

```{r}
total_tested <- bind_rows(replication, rep_fail)
```

"Significant" replication? Split by cis, trans, all

```{r}
total_tested$p[is.na(total_tested$p)] <- 1
total_tested$sigct <- FALSE
total_tested$sigct[total_tested$cis & (total_tested$p < (0.05 / sum(total_tested$cis)))] <- TRUE
total_tested$sigct[!total_tested$cis & (total_tested$p < (0.05 / sum(!total_tested$cis)))] <- TRUE
total_tested$sig <- FALSE
total_tested$sig[total_tested$p < (0.05 / nrow(total_tested))] <- TRUE

replication$sigct <- FALSE
replication$sigct[replication$cis & (replication$p < (0.05 / sum(replication$cis)))] <- TRUE
replication$sigct[!replication$cis & (replication$p < (0.05 / sum(!replication$cis)))] <- TRUE
replication$sig <- FALSE
replication$sig[replication$p < (0.05 / nrow(total_tested))] <- TRUE

total_tested$cist <- "Cis"
total_tested$cist[!total_tested$cis] <- "Trans"
replication$cist <- "Cis"
replication$cist[!replication$cis] <- "Trans"

```

* Cis replication rate at cis threshold: `r sum((total_tested$sigct & total_tested$cis)) / sum(total_tested$cis)`
* Trans replication rate at trans threshold: `r sum((total_tested$sigct & !total_tested$cis)) / sum(!total_tested$cis)`

Higher for trans because the significance threshold is lower. Overall replication rates

* Cis replication rate at overall threshold: `r sum((total_tested$sig & total_tested$cis)) / sum(total_tested$cis)`
* Trans replication rate at overall threshold: `r sum((total_tested$sig & !total_tested$cis)) / sum(!total_tested$cis)`

* Overall replication rate based on overall threshold: `r sum(total_tested$p < (0.05 / n_tested)) / n_tested`

## Effect size concordance


```{r, dev="png"}
ggplot(subset(replication), aes(x=Effect, y=beta1)) +
geom_point(aes(colour=sigct), cex=0.5) +
geom_point(data=subset(replication, !sigct), aes(x = Effect, y = beta1, color = sigct), cex=0.5) +
facet_grid(. ~ cist ) +
labs(x="GoDMC effect size", y="Generation Scotland effect size", colour="Replication") +
theme(legend.position="right") +
scale_fill_brewer(type="qual") +
guides(alpha=FALSE)
ggsave("images/replication.png", width=9, height=4)
ggsave("images/replication.pdf", width=9, height=4)
```

Do smoking associated probes associated explain the bidirectional effects?

```{r, dev="png"}
smok<-read.table("smoking.tsv",he=T,sep="\t")
smok<-subset(smok,PMID=="27651444"& Exposure=="Current versus never smoking") #2622
w<-which(replication$cpg%in%smok$CpG)
replication$smoking<-FALSE
replication$smoking[w]<-TRUE

ggplot(subset(replication), aes(x=Effect, y=beta1)) +
geom_point(aes(colour=sigct),cex=0.5) +
#geom_point(data=subset(replication, !as.logical(sigct)), color="red") +
geom_point(data=subset(replication,sigct == 'FALSE'),aes(x = Effect, y = beta1, color = sigct), cex=0.5) +
geom_point(data=subset(replication,smoking == 'TRUE'),aes(x = Effect, y = beta1), shape=23,cex=0.5) +
facet_grid(. ~ cist ) +
#scale_alpha_manual(values=c(0.02,1)) +
labs(x="GoDMC effect size", y="Generation Scotland effect size", colour="Replication") +
theme(legend.position="right") +
scale_fill_brewer(type="qual") +
guides(alpha=FALSE)
ggsave("images/replication_smok.png", width=9, height=4)
ggsave("images/replication_smok.pdf", width=9, height=4)
```


## Expected vs observed replication rates

The probability of replication is based on the true effect size, the standard errors for replication estimates, and the significance level for replication. Similarly, we can estimate the number of SNPs that should have the same sign. Note that in Ref 21 the notation seems to use the standard error for the discovery to estimate the probability of replication. However, this doesn't make sense to me - if you have an estimate of the true effect size then we expect the replication of that to be related to the expected standard error in the replication sample.

Can we calculate the expected standard error of the replication based on available parameters i.e. effect size, replication MAF, sample size?

$$
r^2 = \frac{t^2}{t^2 + n - 2} \\
2p(1-p)\beta^2 = \frac{\frac{\beta^2}{se^2}}{\frac{\beta^2}{se^2} + n - 2}
$$

rearranging in terms of $s$ gives

$$
s = \sqrt{\frac{
	2p(1 - p)\beta^2 + 1}{
2p(1-p)(n-2)
}}
$$

```{r}
maf <- total_tested$freq1
maf[is.na(maf)] <- total_tested$Freq1[is.na(maf)]
total_tested$expected_rep_se <- sqrt((2 * maf^2 * total_tested$Effect^2 - 2 * maf * total_tested$Effect^2 + 1) / (2 * maf * (-2 + 5101) * (1 - maf)))
```

For a large proportion of discovery estimates tested in the replication we have the known replication standard error. How closely does our expected replication standard error relate to the actual replication standard error?

Correlation R2 = `r cor(total_tested$expected_rep_se, total_tested$se, use="pair")`

```{r}
plot(total_tested$expected_rep_se, total_tested$se, xlab="Expected replication SE", ylab="Actual replication SE")
```

```{r, eval=FALSE}
psig <- function(b, se_disc, alpha)
{
	p_sig <- pnorm(-abs(b) / se_disc + qnorm(alpha / 2)) + (1 - pnorm(-abs(b) / se_disc - qnorm(alpha / 2)))
	return(p_sig)
}

total_tested$prep1 <- psig(total_tested$Effect, total_tested$expected_rep_se, 1e-3)
total_tested$prep2 <- psig(total_tested$Effect, total_tested$expected_rep_se, 0.05/n_tested)

hist(total_tested$prep1, breaks=100)
hist(total_tested$prep2, breaks=100)

ggplot(data=total_tested, aes(x=log(prep1))) +
geom_density(aes(fill=p < 0.05/n_tested))

group_by(total_tested, p < 0.05/n_tested) %>%
do(data.frame(t(quantile(.$prep1, probs = c(0.10, 0.30, 0.50, 0.80)))))

filter(total_tested, p > 0.05/n_tested) %$% preplication(Effect, expected_rep_se, se, 1e-3)



subset(total_tested, !is.na(Effect)) %$% plot(ecdf(psig(Effect,expected_rep_se,1e-3)), type="l")
subset(total_tested, !is.na(Effect)) %$% plot(ecdf(psig(Effect,StdErr,1e-3)), col="red")
subset(total_tested, !is.na(Effect)) %$% plot(ecdf(psig(Effect,StdErrMRE,1e-3)), col="blue")

temp <- subset(total_tested, select=c(p, prep1, prep2)) %>% arrange(prep2)
temp$bin1 <- cut(temp$prep1, 10)
temp$bin2 <- cut(temp$prep2, 10)
group_by(temp, bin1) %>% summarise(n=n(), s=sum(p < 1e-3), p=s/n) %$% plot(bin1, p)
group_by(temp, bin2) %>% summarise(n=n(), s=sum(p < 1e-7), p=s/n) %$% plot(bin2, p)
```




Estimated the expected replication rate across all mQTLs can now be calculated. For comparison, here are the expected replication rates based on the expected SE in the replication; the fixed effects discovery SE; and two random effects discovery SE variables. I think it is more correct to use the expected replication rate of the outcome sample.

```{r}


preplication <- function(b, se_disc, se_rep, alpha)
{
	p_sign <- pnorm(-abs(b) / se_disc) * pnorm(-abs(b) / se_rep) + ((1 - pnorm(-abs(b) / se_disc)) * (1 - pnorm(-abs(b) / se_rep)))
	p_sig <- pnorm(-abs(b) / se_disc + qnorm(alpha / 2)) + (1 - pnorm(-abs(b) / se_disc - qnorm(alpha / 2)))
	message("Total: ", length(b))
	message("Expected sign: ", sum(p_sign, na.rm=TRUE))
	message("Expected sig: ", sum(p_sig, na.rm=TRUE))
	return(list(sum(p_sign, na.rm=TRUE), sum(p_sig, na.rm=TRUE), data_frame(sig=p_sig, sign=p_sign)))
}

preplication_wrapper <- function(total_tested, stderr_col, threshold)
{
	total_tested$STDERR <- total_tested[[stderr_col]]
	x1 <- group_by(total_tested, cist) %>%
	summarise(
		n = n(),
		obs_significant = sum(p < threshold), 
		exp_significant = unlist(preplication(Effect, STDERR, se, threshold)[2]),
		obs_sign = sum(sign(Effect) == sign(beta1), na.rm=TRUE),
		exp_sign = unlist(preplication(Effect, STDERR, se, threshold)[1])
	)
	x2 <- group_by(total_tested) %>%
	summarise(
		n = n(),
		obs_significant = sum(p < threshold), 
		exp_significant = unlist(preplication(Effect, STDERR, se, threshold)[2]),
		obs_sign = sum(sign(Effect) == sign(beta1), na.rm=TRUE),
		exp_sign = unlist(preplication(Effect, STDERR, se, threshold)[1])
	)
	x2$cist <- "all"
	x <- rbind(x1, x2)
	return(x)
}

repl_params <- expand.grid(
	STDERR=c("StdErr", "StdErrMRE", "StdErrARE", "expected_rep_se"),
	threshold = c(1e-3, 1e-4, 1e-5, 1e-6, 0.05/n_tested)
)

l <- list()
for(i in 1:nrow(repl_params))
{
	l[[i]] <- preplication_wrapper(total_tested, as.character(repl_params$STDERR[i]), repl_params$threshold[i])
	l[[i]]$stderr <- repl_params$STDERR[i]
	l[[i]]$beta <- repl_params$BETA[i]
	l[[i]]$threshold <- repl_params$threshold[i]
}
repl_vals <- bind_rows(l)

ggplot(repl_vals, aes(x=-log10(threshold), y=obs_significant/exp_significant)) +
geom_point(aes(colour=cist)) +
geom_line(aes(colour=cist)) +
facet_grid(. ~ stderr) +
scale_colour_brewer(type="qual")

subset(repl_vals, stderr=="expected_rep_se") %>% kable

```

## Strand issues?

```{r}
rep2 <- subset(replication, sigct)
rep2$wrong_dir <- rep2 %$% {sign(Effect) != sign(beta1)}
rep2$allele_codes <- rep2 %$% {paste(Allele1, Allele2)}
```

There are `r sum(rep2$wrong_dir)` mQTLs that have significant replication in the wrong direction. 


Is this due to palindromic SNPs?

```{r}
group_by(rep2, allele_codes) %>%
summarise(
	ncorrect=sum(!wrong_dir), 
	nwrong=sum(wrong_dir), 
	prop=sum(wrong_dir)/n()
) %>% kable
```

There doesn't seem to be a strong relationship with palendromic SNPs. Is there a difference in effect sizes between the replications in the correct vs incorrect directions?

```{r}
group_by(rep2, wrong_dir) %>%
summarise(
	mean_eff = mean(abs(Effect)),
	se_eff = sd(abs(Effect)) / sqrt(n()),
	mean_n = mean(TotalSampleSize),
	mean_isq = mean(HetISq),
	n_studies = mean(HetDf),
	gv = mean(Freq1 * (1-Freq1)),
	prop_cis = sum(cis) / n()
) %>% kable
```

It appears that the effect sizes are on average smaller, which is consistent with them being liable to give the wrong direction. There is also much higher heterogeneity in the discovery, which suggests it could be biological or something to do with the way the data is being treated at those CpGs. What proportion of mQTLs would we expect with wrong sign and significant?


```{r}
r <- replication %$% preplication(beta1, StdErrMRE, se, 0.05/nrow(clumped))[[3]]
exp_wrong_sig <- sum((1 - r$sign) * r$sig)
```

Sum of `(1-prob(correct sign)) * (prob(significant)) = ` `r exp_wrong_sig`

This is a much lower number than observed. Perhaps the SNPs that are leading to these results are more likely to be multi allelic?

Save the replication data

```{r}
save(replication, file="replication.rdata")
```