---
title: Replication of GoDMC results in Generation Scotland
output:
  html_document:
    toc: true
    theme: united
---

```{r, echo=FALSE}
suppressWarnings(suppressPackageStartupMessages(library(dplyr)))
suppressWarnings(suppressPackageStartupMessages(library(magrittr)))
suppressWarnings(suppressPackageStartupMessages(library(data.table)))
suppressWarnings(suppressPackageStartupMessages(library(ggplot2)))
suppressWarnings(suppressPackageStartupMessages(library(tidyr)))
suppressWarnings(suppressPackageStartupMessages(library(knitr)))
suppressWarnings(suppressPackageStartupMessages(library(scales)))
opts_chunk$set(cache=TRUE, echo=FALSE, message=FALSE, warning=FALSE)
load("replication.rdata")
cis<-replication[which(replication$cist=="Cis"),]
trans<-replication[which(replication$cist=="Trans"),]

```
There are `r nrow(replication)` associations in the replication file of which `r nrow(cis)` in cis and `r nrow(trans)`

## Variance explained

Using an independent replication allows unbiased estimate of phenotypic variance explained

```{r, echo=FALSE}

ggplot(replication, aes(x=r2)) +
geom_density(aes(fill=cist), alpha=0.3) +
labs(x="DNA methylation variance explained", fill="mQTL")
ggsave("images/varexp.pdf")

ggplot(replication, aes(x = r2,color=cist)) +  
geom_bar(aes(y = (..count..)/sum(..count..))) + 
scale_y_continuous(labels=percent)
ggsave("images/varexp_histperc.pdf")

ggplot(replication, aes(x=r2,fill=cist),color=cist,alpha=0.3) +
geom_histogram(position="dodge") +
labs(x="DNA methylation variance explained",color="mQTL",fill="mQTL")
ggsave("images/varexp_hist.pdf")

ggplot(replication, aes(x=r2,color=cist),alpha=0.3) +
stat_ecdf() +
scale_x_continuous(breaks=seq(0,1,0.05)) +
labs(x="DNA methylation variance explained", y="Proportion of DNAm sites",color="mQTL")

ggsave("images/varcumulative_exp.pdf")

y <- cumsum(table(round(cis$r2,4))/nrow(cis))
y<-data.frame(names(y),y)
cis99<-min(as.numeric(as.character(y[y[,2]>0.99,1])))

y <- cumsum(table(round(trans$r2,4))/nrow(trans))
y<-data.frame(names(y),y)
trans99<-min(as.numeric(as.character(y[y[,2]>0.99,1])))


```

---

Total phenotypic variance explained by all mQTLs:

```{r}
replication$r2 / 420509
```

By Cis only:

```{r}
group_by(replication, cist) %>%
summarise(vpe = sum(r2)/420509)
```

The cumulative plot shows variance explained in various percentages of DNA methylation sites. 99 % of the cis associations explain less than `r cis99` and 99% of the trans associations explain less than `r trans99` of the variance.


```{r, echo=FALSE}
replication2 <- group_by(replication, cpg) %>%
	summarise(r2=sum(r2))
replication2$what <- "hsqtot"

cis2 <- group_by(cis, cpg) %>%
	summarise(r2=sum(r2))
cis2$what <- "hsqtot"

trans2 <- group_by(trans, cpg) %>%
	summarise(r2=sum(r2))
trans2$what <- "hsqtot"

load("~/repo/godmc_phase2_analysis/07_enrichments/mean_allcpgs.Robj")
m<-match(replication2$cpg,df.all$cpg)
replication_mean_sd<-data.frame(replication2,df.all[m,c("meancpg","sdcpg")])

breaks <- seq(0,1,0.1)
# specify interval/bin labels
labels <- c("0-10%", "10-20%", "20-30%", "30-40%", "40-50%", "50-60%", "60-70%", "70-80%","80-90%","90-100%")
# bucketing data points into bins
bins <- cut(replication_mean_sd$meancpg, breaks, include.lowest = T, right=FALSE, labels=labels)

breaks2 <- seq(0,0.125,0.025)
labels2<-c("0-20%", "20-40%", "40-60%", "60-80%","80-100%")
bins2 <- cut(replication_mean_sd$sdcpg, breaks2, include.lowest = T, right=FALSE, labels=labels2)

breaks3 <- c(0,0.2,0.8,1.0)
labels3<-c("0-20%", "20-80%","80-100%")
bins3 <- cut(replication_mean_sd$meancpg, breaks3, include.lowest = T, right=FALSE, labels=labels3)

replication_mean_sd<-data.frame(replication_mean_sd,mean.bins=bins3,sd.bins=bins3)

p1<-ggplot(replication_mean_sd,aes(x=mean.bins,y=r2)) +
geom_violin() +
geom_boxplot(width=0.1,fill="white") +
labs(x="Weighted mean DNA methylation level",y="variance explained (r2)")
ggsave(p1,file="images/meanbeta_r2.pdf")

ggplot(replication_mean_sd,aes(x=sd.bins,y=r2)) +
geom_boxplot() +
labs(x="Weighted sd DNA methylation level",y="variance explained (r2)")
ggsave("images/meansd_r2.pdf")

replication_mean_sd %>% lm(r2 ~ meancpg, data=.) %>% summary %>% coefficients
replication_mean_sd %>% lm(r2 ~ sdcpg, data=.) %>% summary %>% coefficients


```

This plot shows that sites with intermediate methylation have larger r2
 

We previously calculated (with low precision) the SNP h2 of each CpG

```{r, echo=FALSE}
load("/panfs/panasas01/shared-godmc/heritability/hsq_ct_aries.RData")
dat$cpg <- as.character(dat$CPG)
dat <- subset(dat, timepoint == "FOM", select=c(hsq1, hsq2, cpg))
dat$hsqtot <- dat$hsq1 + dat$hsq2
dat <- gather(dat, key=what, val=hsq, -cpg)

#dat <- gather(dat, val=hsq, -cpg)
#replication2 <- group_by(replication, cpg) %>%
#	summarise(r2=sum(r2))
#replication2$what <- "hsqtot"

replication3 <- group_by(replication, cpg, cis) %>%
	summarise(r2=sum(r2))
replication3$what <- "hsq1"
replication3$what[!replication3$cis] <- "hsq2"
replication3 <- subset(replication3, select=-c(cis))

group_by(replication3, what) %>% summarise(m=mean(r2), mf={a = 420509-n(); mean(c(r2, rep(0,a)), na.rm=T)})
group_by(replication2) %>% summarise(m=mean(r2), mf={a = 420509-nrow(replication2); mean(c(r2, rep(0,a)), na.rm=T)})
	
replication4 <- bind_rows(replication3, replication2)
replication4 <- merge(dat, replication4, by=c("cpg", "what"), all.x=TRUE)
replication4$r2[is.na(replication4$r2)] <- 0

out <- group_by(replication4, what) %>%
	summarise(mean = mean(hsq-r2), median = median(hsq-r2))
out$what <- c("cis", "trans", "total")
# kable(out)
```


How does the SNP h2 (estimated in ARIES) relate to the variance explained by clumped SNPs (Generation Scotland)

```{r, echo=FALSE, eval=TRUE, dev="png"}
replication5 <- subset(replication4, what=="hsqtot")
replication5$ord <- rank(replication5$hsq)
replication5 <- gather(subset(replication5, select=c(hsq, r2, ord)), key=what, value=value, -ord)
replication5$what[replication5$what == "r2"] <- "gs_r2"
replication5 <- replication5[nrow(replication5):1, ]
ggplot(replication5, aes(x=ord, y=value)) +
geom_point(aes(colour=what)) +
geom_smooth(data=subset(replication5, what=="gs_r2"))
```


```{r, echo=FALSE, eval=FALSE, dev="png"}
replication5 <- group_by(replication4, what) %>%
	mutate(ord = rank(hsq)/n()) 

replication5 <- subset(replication5, select=c(what, hsq, r2, ord)) %>%
	group_by(what) %>%
	do({
		gather(., key=key, value=value, -ord)
	})
replication5 <- replication5[nrow(replication5):1, ]
ggplot(replication5[sample(1:nrow(replication5), 100000),], aes(x=ord, y=value)) +
geom_point(aes(colour=key)) +
geom_smooth(data=subset(replication5, what=="gs_r2")) +
facet_grid(what ~ .)
```

Use Van Dongen heritability estimates using twins

```{r, echo=FALSE}
# load("vanDongen_h2gt50_probes.RData")
vandongen <- read.csv("/panfs/panasas01/shared-godmc/heritability/vanDongen_HeritabilityofDNAmethylation.csv")
p <- sum(vandongen$cgid %in% replication2$cpg)

```
* There are `r nrow(vandongen)` CpGs with significant h2 estimates in van Dongen. 
* `r p` were present in the GS replication dataset

What proportion of the h2 is explained by mQTLs?

```{r, echo=FALSE}
vd <- inner_join(replication2, vandongen, by=c("cpg"="cgid"))
vd$ord <- rank(vd$h2_twinAE)

vd %>% filter(h2_twinAE != 0) %>%
	group_by()

w0<-which(vd$h2_twinAE == 0)

```
The mean h2 is `r mean(vd$h2_twinAE,na.rm=T)` across `r nrow(vd)` sites. 
Correlation between variance explained in GoDMC and h2_twin incl h2==0: `r cor(vd$r2, vd$h2_twinAE)`
Correlation between variance explained in GoDMC and h2_twin incl h2!=0: `r cor(vd$r2[-w0], vd$h2_twinAE[-w0])`
```{r, echo=FALSE}

vd %>% lm(r2 ~ h2_twinAE, data=.) %>% summary %>% coefficients

ggplot(subset(vd, !is.na(h2_twinAE) & !is.na(r2)), aes(x=as.factor(h2_twinAE), y=r2)) +
geom_boxplot(outlier.size=0.01) +
theme(axis.text.x=element_text(angle=90, size=3)) +
labs(x="Twin heritability estimate", y="Variance explained in GS replication")
```

What proportion of h2 is explained? To summarise this set probes with h2==0 to h2==1e-10 because this gives division errors. r2 for probes with h2==0 are set to 0.

* There are `r length(which(vd$h2_twinAE == 0))` CpGs with h2==0.

```{r, echo=FALSE}
vd2 <- gather(subset(vd, h2_twinAE != 0, select=c(r2, h2_twinAE, ord)),key, value, -ord)

vd2<-subset(vd, h2_twinAE != 0)
vd2$prop <- (vd2$r2 / vd2$h2_twinAE)
w<-which(vd2$prop>1)
vd2$prop[w]<-1


vd$h2_twinAE[w0]<-1e-10
vd$prop <- (vd$r2 / vd$h2_twinAE)
vd$prop[w0]<-0

ggplot(vd, aes(x = prop)) +  
geom_histogram()
ggsave(file="images/prop_vandongen.pdf")
```

There are `r length(which(vd$prop>1))` proportions above 1. These will be set to 1.

```{r, echo=FALSE}
w<-which(vd$prop>1)
vd$prop[w]<-1

ggplot(vd, aes(x = prop)) +  
geom_histogram()
ggsave(file="images/prop_vandongen_all.pdf")


ggplot(vd[-w0,], aes(x = prop)) +  
geom_histogram()
ggsave(file="images/prop_vandongen_no_h2eq0.pdf")

```

What proportion of h2 is explained? h2==0 has been replaced with 1e-10

```{r, echo=FALSE}

vd3 <- gather(subset(vd, select=c(r2, h2_twinAE, prop, ord)),key, value, -ord)

	group_by(vd3, key) %>%
	summarise(mean=mean(value), median=median(value)) %>%
	kable
```

What proportion of h2 is explained? h2==0 has been removed

```{r, echo=FALSE}

vd4<-gather(subset(vd2, select=c(r2, h2_twinAE, prop, ord)),key, value, -ord)
    group_by(vd4, key) %>%
	summarise(mean=mean(value), median=median(value)) %>%
	kable

```

Do the same with McRae pedigree-based h2 estimates

```{r, echo=FALSE}
# load("vanDongen_h2gt50_probes.RData")
mcrae <- fread("https://static-content.springer.com/esm/art%3A10.1186%2Fgb-2014-15-5-r73/MediaObjects/13059_2013_3271_MOESM1_ESM.txt")
names(mcrae) <- c("cpg", "chr", "pos", "ncpg", "probetype", "classification", "nsnp", "h2", "h2_cc")
p <- sum(mcrae$cpg %in% replication2$cpg)
```

* There are `r nrow(mcrae)` CpGs with significant h2 estimates in McRae. 
* `r p` were present in the GS replication dataset

What proportion of the h2 is explained by mQTLs?

```{r, echo=FALSE}
mc <- inner_join(replication2, mcrae, by=c("cpg"="cpg"))
mc$ord <- rank(mc$h2_cc)
mc$h2_cc_r <- round(mc$h2_cc, 2)

w0<-which(mc$h2_cc == 0)

```

```{r, echo=FALSE}
ggplot(subset(mc, !is.na(h2_cc) & !is.na(r2)), aes(x=as.factor(h2_cc_r), y=r2)) +
geom_boxplot(outlier.size=0.01) +
theme(axis.text.x=element_text(angle=90, size=3)) +
labs(x="Pedigree heritability estimate", y="Variance explained in GS replication")

mc%>% lm(r2 ~ h2_cc, data=.) %>% summary %>% coefficients
```

The mean h2 is `r mean(mc$h2_cc,na.rm=T)` across `r nrow(mc)` sites. 
Correlation between variance explained in GoDMC and h2_pedigree: `r cor(mc$r2, mc$h2_cc)`
Correlation between variance explained in GoDMC and h2_pedigree: `r cor(mc$r2[-w0], mc$h2_cc[-w0])`

What proportion of h2 is explained? To summarise this probes with h2==0 are replaced with h2==1e-10 to avoid division errors.

* There are `r length(which(mc$h2_cc == 0))` CpGs with h2==0.


```{r, echo=FALSE}
mc2 <- gather(subset(mc, h2_cc != 0, select=c(r2, h2_cc, ord)),key, value, -ord)
mc2 <- subset(mc, h2_cc != 0)
mc2$prop <-(mc2$r2 / mc2$h2_cc)
w<-which(mc2$prop>1)
mc2$prop[w]<-1

mc$h2_cc[w0]<-1e-10
mc$prop <- (mc$r2 / mc$h2_cc)
mc$prop[w0]<-0

```
There are `r length(which(mc$prop>1))` proportions above 1. These will be set to 1.

```{r, echo=FALSE}
w<-which(mc$prop>1)
mc$prop[w]<-1

ggplot(mc, aes(x = prop)) +  
geom_histogram()
ggsave(file="images/prop_mcrae.pdf")

ggplot(mc2, aes(x = prop)) +  
geom_histogram()
ggsave(file="images/prop_mcrae_no_h2eq0.pdf")

mc3 <- gather(subset(mc, select=c(r2, h2_cc, prop, ord)),key, value, -ord)

	group_by(mc3, key) %>%
	summarise(mean=mean(value), median=median(value)) %>%
	kable

mc4 <- gather(subset(mc2, select=c(r2, h2_cc, prop, ord)),key, value, -ord)

	group_by(mc4, key) %>%
	summarise(mean=mean(value), median=median(value)) %>%
	kable
```

## Visualise proportion variance explained

```{r}
mc5 <- group_by(subset(mc, h2_cc_r!=0), h2_cc_r) %>%
summarise(n=n(), r2=mean(r2), propm=mean(prop), propsd=sd(prop))

mc5$propb <- pmin(mc5$propm, 1)

range01 <- function(x) pmax(pmin(x, 1), 0)

ggplot(mc5, aes(x=h2_cc_r, y=propb)) +
geom_errorbar(aes(ymin=range01(propb-propsd), ymax=range01(propm+propsd))) +
geom_point(aes(size=n, colour=n)) +
# ylim(0, 1) +
labs(x="Estimated heritability (binned)", y="Proportion of genetic variance explained", size="Number of DNAm sites", colour="Number of DNAm sites") +
scale_color_continuous(limits=c(0, 2500), breaks=seq(0, 2500, by=500)) +
scale_size_continuous(limits=c(0, 2500), breaks=seq(0, 2500, by=500)) +
guides(color= guide_legend(), size=guide_legend())
ggsave("images/prop_h2_explained.pdf",height=6,width=10)

##
mc6 <- group_by(mc, h2_cc_r) %>%
summarise(n=n(), r2=mean(r2), propm=mean(prop), propsd=sd(prop))

mc6$propb <- pmin(mc6$propm, 1)

range01 <- function(x) pmax(pmin(x, 1), 0)

ggplot(mc6, aes(x=h2_cc_r, y=propb)) +
geom_errorbar(aes(ymin=range01(propb-propsd), ymax=range01(propm+propsd))) +
geom_point(aes(size=n, colour=n)) +
# ylim(0, 1) +
labs(x="Estimated heritability (binned)", y="Proportion of genetic variance explained", size="Number of DNAm sites", colour="Number of DNAm sites") +
scale_color_continuous(limits=c(0, 7500), breaks=seq(0, 7500, by=1000)) +
scale_size_continuous(limits=c(0, 7500), breaks=seq(0, 7500, by=1000)) +
guides(color= guide_legend(), size=guide_legend())
ggsave("images/prop_h2_explained_incl_h2eq0.pdf",height=6,width=10)
```

## Create tables

```{r}
# Replication

tab1 <- with(replication, data_frame(
	snp=snp, cpg=cpg, effect_allele=Allele1, other_allele=Allele2, discovery_beta=Effect, discovery_se=StdErr, discovery_n=TotalSampleSize, discovery_pval=pval, discovery_isq=HetISq, cis=cis,
	replication_beta=beta1, replication_se=se, replication_n=n, replication_pval=p, replication_rsq=r2
))

# h2

tab2a <- with(mcrae, data_frame(
	cpg=cpg, h2_mcrae=h2_cc
))
nrow(tab2a)

tab2b <- with(vandongen, data_frame(
	cpg=cgid, h2_vandongen=h2_twinAE
))
nrow(tab2b)

#tab2 <- inner_join(tab2a, tab2b, by="cpg")
tab2 <- full_join(tab2a, tab2b, by="cpg")
nrow(tab2)

names(cis2)<-gsub("r2","r2.cis",names(cis2))
names(trans2)<-gsub("r2","r2.trans",names(trans2))

tab2 <- merge(tab2, subset(replication2, select=-c(what)), by="cpg", all.x=TRUE)
tab2 <- merge(tab2, subset(cis2, select=-c(what)), by="cpg", all.x=TRUE)
tab2 <- merge(tab2, subset(trans2, select=-c(what)), by="cpg", all.x=TRUE)


tab2<-tab2[which(tab2$cpg%in%df.all$cpg),]
nrow(tab2)

gs <- read.csv("sites_to_exclude_5200_samples.csv")
tab2<-tab2[which(tab2$cpg%in%gs$x==F),]
nrow(tab2)
length(which(is.na(tab2$r2))) #277428

tab2$r2[is.na(tab2$r2)] <- 0
tab2$r2.cis[is.na(tab2$r2.cis)] <- 0
tab2$r2.trans[is.na(tab2$r2.trans)] <- 0

load("../results/16/16_clumped.rdata")
clumped<-clumped[which(clumped$cis==TRUE & clumped$pval<1e-8 | clumped$cis==FALSE & clumped$pval<1e-14),]
mqtlcount <- group_by(clumped, cpg) %>% summarise(mqtls=n())
clumped.cis<-clumped[which(clumped$cis==TRUE & clumped$pval<1e-8),]
clumped.trans<-clumped[which(clumped$cis==FALSE & clumped$pval<1e-14),]

ciscount<-group_by(clumped.cis, cpg) %>% summarise(cismqtls=n())
transcount<-group_by(clumped.trans, cpg) %>% summarise(transmqtls=n())

head(clumped)
o<-order(abs(clumped$Effect),decreasing=T)
clumped<-clumped[o,]
clumped[which(clumped$cpg%in%c("cg03228974")),]

m<-match(tab2$cpg,clumped$cpg)
tab2<-data.frame(tab2,clumped[m,"Effect"])
tab2[which(tab2$cpg%in%c("cg03228974")),]


tab2l <- tidyr::gather(tab2, key="key", value="value", h2_mcrae, h2_vandongen)
tab2l$key[tab2l$key=="h2_mcrae"] <- "Twin family"
tab2l$key[tab2l$key=="h2_vandongen"] <- "Twin"


w<-which(tab2l$value==0)
length(w)
table(tab2l$key[w])
tab2l$r2[w]<-0

w.miss<-which(is.na(tab2l$value))
table(tab2l$key[w.miss])

w<-which(is.na(tab2l$Effect))
length(w)

ggplot(tab2l[-w.miss,], aes(x=as.factor(round(value*2, 1)/2), y=abs(Effect))) +
geom_boxplot(outlier.size=0.01, aes(fill=key)) +
theme(axis.text.x=element_text(angle=90, size=3)) +
labs(x="Heritability estimate", y="Absolute Maximum Effect Size", fill="Heritability\nstudy design")
ggsave(file="images/h2_vs_effectsize.png", width=9, height=4)
ggsave(file="images/h2_vs_effectsize.pdf", width=9, height=4)


ggplot(tab2l, aes(x=as.factor(round(value*2, 1)/2), y=r2)) +
geom_boxplot(outlier.size=0.01, aes(fill=key)) +
theme(axis.text.x=element_text(angle=90, size=3)) +
labs(x="Heritability estimate", y="Variance explained in GS replication", fill="Heritability\nstudy design")
ggsave(file="images/h2_vs_r2.png", width=9, height=4)
ggsave(file="images/h2_vs_r2.pdf", width=9, height=4)


tab2l <- merge(tab2l, mqtlcount, by="cpg", all.x=TRUE)
tab2l$mqtls[is.na(tab2l$mqtls)] <- 0

tab2l <- merge(tab2l, ciscount, by="cpg", all.x=TRUE)
tab2l$cismqtls[is.na(tab2l$cismqtls)] <- 0
tab2l <- merge(tab2l, transcount, by="cpg", all.x=TRUE)
tab2l$transmqtls[is.na(tab2l$transmqtls)] <- 0

ggplot(tab2l, aes(x=as.factor(mqtls), y=value)) +
geom_boxplot(outlier.size=0.01, aes(fill=key)) +
theme(axis.text.x=element_text(angle=90)) +
labs(x="Number of mQTL per DNA methylation site", y="Heritability estimate", fill="Heritability\nstudy design")
ggsave(file="images/mqtl_vs_h2.png", width=9, height=4)
ggsave(file="images/mqtl_vs_h2.pdf", width=9, height=4)

tab3l<-tab2l[which(tab2l$key=="Twin family"),]
tab3l$r2.diff<-(tab3l$r2.cis-tab3l$r2.trans)
w<-which(tab3l$r2.diff<0)
length(w) #6629
tr<-abs(tab3l[w,"r2.diff"])
length(which(tr>0.25)) #33
length(which(tr>0.10)) #412

ggplot(tab3l, aes(x=r2.diff)) +
geom_histogram() +
theme(axis.text.x=element_text(angle=90)) +
labs(x="r2 difference (cis-trans)")
ggsave(file="images/r2difference_cis_trans.png", width=9, height=4)

tab3l <- tidyr::gather(tab3l, key="key2", value="cistrans.r2", r2.cis, r2.trans)
tab3l$key2[tab3l$key2=="r2.cis"] <- "cis"
tab3l$key2[tab3l$key2=="r2.trans"] <- "trans"

w<-which(tab3l$key2=="cis")
tab3l$cistransmqtls[w]<-tab3l$cismqtls[w]
w<-which(tab3l$key2=="trans")
tab3l$cistransmqtls[w]<-tab3l$transmqtls[w]



ggplot(tab3l, aes(x=as.factor(mqtls), y=r2)) +
geom_boxplot(outlier.size=0.01) +
theme(axis.text.x=element_text(angle=90)) +
labs(x="Number of mQTL per DNA methylation site", y="Variance explained in GS replication")
ggsave(file="images/mqtl_vs_r2.png", width=9, height=4)
ggsave(file="images/mqtl_vs_r2.pdf", width=9, height=4)

p1<-ggplot(tab4l, aes(x=as.factor(cistransmqtls), y=cistrans.r2)) +
geom_boxplot(outlier.size=0.01, aes(fill=key2)) +
theme(axis.text.x=element_text(angle=90)) +
labs(x="Number of mQTL per DNA methylation site", y="Variance explained in GS replication", fill="Annotation")
ggsave(p1,file="images/mqtl_vs_r2_cistrans.png", width=9, height=4)
ggsave(file="images/mqtl_vs_r2_cistrans.pdf", width=9, height=4)


```

regression of mqtls~h2_McRae

```{r}
tab2l %>% filter(key == "Family") %>% lm(mqtls ~ value, data=.) %>% summary %>% coefficients
```

correlation mqtls~h2_McRae

```{r}
#tab2l %>% group_by(key) %>% summarise(cor=cor(mqtls, value,use="complete.obs")^2)
tab2l %>% group_by(key) %>% summarise(cor=cor(mqtls, value,use="complete.obs"))
```
regression of r2~h2_twin using `r nrow(tab2l[tab2l$key=="Twin",])` probes, missing r2 are set to 0

```{r}
tab2l %>% filter(key == "Twin") %>% lm(value ~ r2, data=.) %>% summary %>% coefficients
```

regression of h2_McRae~r2 using `r nrow(tab2l[tab2l$key=="Twin family",])` probes, missing r2 are set to 0
regression of h2_McRae~r2 using `r nrow(tab2l[tab2l$key=="Twin family",])` probes, missing r2 are set to 0

```{r}
tab2l %>% filter(key == "Family") %>% lm(value ~ r2, data=.) %>% summary %>% coefficients


ggplot(tab2l, aes(x=as.factor(round(r2*4, 1)/4), y=value)) +
geom_boxplot(outlier.size=0.01, aes(fill=key)) +
theme(axis.text.x=element_text(angle=90)) +
labs(x="Variance explained in GS replication", y="Heritability estimate", fill="Heritability\nstudy design")
ggsave(file="images/r2_vs_h2.png", width=9, height=4)
ggsave(file="images/r2_vs_h2.pdf", width=9, height=4)

write.csv(tab1, file="results/replication.csv")

#tab3<-tab2[which(tab2$cpg %in% replication2$cpg),]
tab3<-left_join(replication2[,1],tab2,by="cpg")

tab3$genetic_r2_mcrae<-tab3$r2/tab3$h2_mcrae
tab3$genetic_r2_vandongen<-tab3$r2/tab3$h2_vandongen

w1<-which(tab3$h2_mcrae==0)
tab3$genetic_r2_mcrae[w1]<-0
w2<-which(tab3$h2_vandongen==0)
tab3$genetic_r2_vandongen[w2]<-0

w1<-which(tab3$genetic_r2_mcrae>1)
tab3$genetic_r2_mcrae[w1]<-1
w2<-which(tab3$h2_vandongen>1)
tab3$genetic_r2_vandongen[w2]<-1

write.csv(tab3, file="results/variance_explained.csv")

```





