---
title: Replication of GoDMC results in Generation Scotland
output:
  html_document:
    toc: true
    theme: united
---

```{r, echo=FALSE}
suppressWarnings(suppressPackageStartupMessages(library(dplyr)))
suppressWarnings(suppressPackageStartupMessages(library(magrittr)))
suppressWarnings(suppressPackageStartupMessages(library(data.table)))
suppressWarnings(suppressPackageStartupMessages(library(ggplot2)))
suppressWarnings(suppressPackageStartupMessages(library(tidyr)))
suppressWarnings(suppressPackageStartupMessages(library(knitr)))
opts_chunk$set(cache=TRUE, echo=FALSE, message=FALSE, warning=FALSE)
load("replication.rdata")
```

## Variance explained

Using an independent replication allows unbiased estimate of phenotypic variance explained

```{r, echo=FALSE}

ggplot(replication, aes(x=r2)) +
geom_density(aes(fill=cist), alpha=0.3) +
labs(x="DNA methylation variance explained", fill="mQTL")
ggsave("images/varexp2.pdf")

ggplot(replication, aes(x=r2,color=cist),alpha=0.3) +
stat_ecdf() +
scale_x_continuous(breaks=seq(0,1,0.05)) +
labs(x="DNA methylation variance explained", y="Proportion of DNA methylation sites",color="mQTL")
ggsave("images/varcumulative_exp2.pdf")
```

We previously calculated (with low precision) the SNP h2 of each CpG

```{r, echo=FALSE}
load("/panfs/panasas01/shared-godmc/heritability/hsq_ct_aries.RData")
dat$cpg <- as.character(dat$CPG)
dat <- subset(dat, timepoint == "FOM", select=c(hsq1, hsq2, cpg))
dat$hsqtot <- dat$hsq1 + dat$hsq2
dat <- gather(dat, key=what, val=hsq, -cpg)

#dat <- gather(dat, val=hsq, -cpg)
replication2 <- group_by(replication, cpg) %>%
	summarise(r2=sum(r2))
replication2$what <- "hsqtot"

replication3 <- group_by(replication, cpg, cis) %>%
	summarise(r2=sum(r2))
replication3$what <- "hsq1"
replication3$what[!replication3$cis] <- "hsq2"
replication3 <- subset(replication3, select=-c(cis))

group_by(replication3, what) %>% summarise(m=mean(r2), mf={a = 420509-n(); mean(c(r2, rep(0,a)), na.rm=T)})
group_by(replication2) %>% summarise(m=mean(r2), mf={a = 420509-nrow(replication2); mean(c(r2, rep(0,a)), na.rm=T)})
	

replication4 <- bind_rows(replication3, replication2)
replication4 <- merge(dat, replication4, by=c("cpg", "what"), all.x=TRUE)
replication4$r2[is.na(replication4$r2)] <- 0

out <- group_by(replication4, what) %>%
	summarise(mean = mean(hsq-r2), median = median(hsq-r2))
out$what <- c("cis", "trans", "total")
# kable(out)
```


How does the SNP h2 (estimated in ARIES) relate to the variance explained by clumped SNPs (Generation Scotland)

```{r, echo=FALSE, eval=TRUE, dev="png"}
replication5 <- subset(replication4, what=="hsqtot")
replication5$ord <- rank(replication5$hsq)
replication5 <- gather(subset(replication5, select=c(hsq, r2, ord)), key=what, value=value, -ord)
replication5$what[replication5$what == "r2"] <- "gs_r2"
replication5 <- replication5[nrow(replication5):1, ]
ggplot(replication5, aes(x=ord, y=value)) +
geom_point(aes(colour=what)) +
geom_smooth(data=subset(replication5, what=="gs_r2"))
```


```{r, echo=FALSE, eval=FALSE, dev="png"}
replication5 <- group_by(replication4, what) %>%
	mutate(ord = rank(hsq)/n()) 

replication5 <- subset(replication5, select=c(what, hsq, r2, ord)) %>%
	group_by(what) %>%
	do({
		gather(., key=key, value=value, -ord)
	})
replication5 <- replication5[nrow(replication5):1, ]
ggplot(replication5[sample(1:nrow(replication5), 100000),], aes(x=ord, y=value)) +
geom_point(aes(colour=key)) +
geom_smooth(data=subset(replication5, what=="gs_r2")) +
facet_grid(what ~ .)
```

Use Van Dongen heritability estimates using twins

```{r, echo=FALSE}
# load("vanDongen_h2gt50_probes.RData")
vandongen <- read.csv("/panfs/panasas01/shared-godmc/heritability/vanDongen_HeritabilityofDNAmethylation.csv")
p <- sum(vandongen$cgid %in% replication2$cpg)
```

* There are `r nrow(vandongen)` CpGs with significant h2 estimates in van Dongen. 
* `r p` were present in the GS replication dataset

What proportion of the h2 is explained by mQTLs?

```{r, echo=FALSE}
vd <- inner_join(replication2, vandongen, by=c("cpg"="cgid"))
vd$ord <- rank(vd$h2_twinAE)

vd %>% filter(h2_twinAE != 0) %>%
	group_by()

```

```{r, echo=FALSE}
ggplot(subset(vd, !is.na(h2_twinAE) & !is.na(r2)), aes(x=as.factor(h2_twinAE), y=r2)) +
geom_boxplot(outlier.size=0.01) +
theme(axis.text.x=element_text(angle=90, size=3)) +
labs(x="Twin heritability estimate", y="Variance explained in GS replication")
```

```{r, echo=FALSE}
vd2 <- gather(subset(vd, h2_twinAE != 0, select=c(r2, h2_twinAE, ord)),key, value, -ord)
vd <- subset(vd, h2_twinAE != 0)
vd$prop <- (vd$r2 / vd$h2_twinAE)
vd3 <- gather(subset(vd, select=c(r2, h2_twinAE, prop, ord)),key, value, -ord)

	group_by(vd3, key) %>%
	summarise(mean=mean(value), median=median(value)) %>%
	kable
```

Correlation between variance explained in GoDMC and h2_twin: `r cor(vd$r2, vd$h2_twinAE)`

Do the same with McRae pedigree-based h2 estimates

```{r, echo=FALSE}
# load("vanDongen_h2gt50_probes.RData")
mcrae <- fread("https://static-content.springer.com/esm/art%3A10.1186%2Fgb-2014-15-5-r73/MediaObjects/13059_2013_3271_MOESM1_ESM.txt")
names(mcrae) <- c("cpg", "chr", "pos", "ncpg", "probetype", "classification", "nsnp", "h2", "h2_cc")
p <- sum(mcrae$cpg %in% replication2$cpg)
```

* There are `r nrow(mcrae)` CpGs with significant h2 estimates in McRae. 
* `r p` were present in the GS replication dataset

What proportion of the h2 is explained by mQTLs?

```{r, echo=FALSE}
mc <- inner_join(replication2, mcrae, by=c("cpg"="cpg"))
mc$ord <- rank(mc$h2_cc)
mc$h2_cc_r <- round(mc$h2_cc, 2)
```

```{r, echo=FALSE}
ggplot(subset(mc, !is.na(h2_cc) & !is.na(r2)), aes(x=as.factor(h2_cc_r), y=r2)) +
geom_boxplot(outlier.size=0.01) +
theme(axis.text.x=element_text(angle=90, size=3)) +
labs(x="Pedigree heritability estimate", y="Variance explained in GS replication")

```

```{r, echo=FALSE}
mc2 <- gather(subset(mc, h2_cc != 0, select=c(r2, h2_cc, ord)),key, value, -ord)
mc <- subset(mc, h2_cc != 0)
mc$prop <-(mc$r2 / mc$h2_cc)
mc3 <- gather(subset(mc, select=c(r2, h2_cc, prop, ord)),key, value, -ord)

	group_by(mc3, key) %>%
	summarise(mean=mean(value), median=median(value)) %>%
	kable
```

Correlation between variance explained in GoDMC and h2_pedigree: `r cor(mc$r2, mc$h2_cc)`

Visualise proportion variance explained

```{r}
mc4 <- group_by(subset(mc, h2_cc_r!=0), h2_cc_r) %>%
summarise(n=n(), r2=mean(r2), propm=mean(prop), propsd=sd(prop))
mc4$propb <- pmin(mc4$propm, 1)

range01 <- function(x) pmax(pmin(x, 1), 0)

ggplot(mc4, aes(x=h2_cc_r, y=propb)) +
geom_errorbar(aes(ymin=range01(propb-propsd), ymax=range01(propm+propsd))) +
geom_point(aes(size=n, colour=n)) +
# ylim(0, 1) +
labs(x="Estimated heritability (binned)", y="Proportion of variance explained", size="Number of CpGs in bin", colour="Number of CpGs in bin") +
scale_color_continuous(limits=c(0, 2500), breaks=seq(0, 2500, by=500)) +
scale_size_continuous(limits=c(0, 2500), breaks=seq(0, 2500, by=500)) +
guides(color= guide_legend(), size=guide_legend())
ggsave("images/prop_h2_explained.pdf")
```

## Create tables

```{r}
# Replication

tab1 <- with(replication, data_frame(
	snp=snp, cpg=cpg, effect_allele=Allele1, other_allele=Allele2, discovery_beta=Effect, discovery_se=StdErr, discovery_n=TotalSampleSize, discovery_pval=pval, discovery_isq=HetISq, cis=cis,
	replication_beta=beta1, replication_se=se, replication_n=n, replication_pval=p, replication_rsq=r2
))

# h2

tab2a <- with(mcrae, data_frame(
	cpg=cpg, h2_mcrae=h2_cc
))

tab2b <- with(vandongen, data_frame(
	cpg=cgid, h2_vandongen=h2_twinAE
))

tab2 <- inner_join(tab2a, tab2b, by="cpg")
tab2 <- merge(tab2, subset(replication2, select=-c(what)), by="cpg", all.x=TRUE)
tab2$r2[is.na(tab2$r2)] <- 0

tab2l <- tidyr::gather(tab2, key="key", value="value", h2_mcrae, h2_vandongen)
tab2l$key[tab2l$key=="h2_mcrae"] <- "Family"
tab2l$key[tab2l$key=="h2_vandongen"] <- "Twin"

p1 <- ggplot(tab2l, aes(x=as.factor(round(value*2, 1)/2), y=r2)) +
geom_boxplot(outlier.size=0.01, aes(fill=key)) +
theme(axis.text.x=element_text(angle=90, size=3)) +
labs(x="Heritability estimate", y="Variance explained in GS replication", fill="Heritability\nstudy design")
ggsave(p1, file="images/h2_vs_r2.png", width=9, height=4)

load("../results/16/16_clumped.rdata")
clumped<- subset(clumped, pval < 1e-8 & cis | pval < 1e-14 & !cis)

clumped$cist <- "cis"
clumped$cist[!clumped$cis] <- "trans"
clumped <- subset(clumped, snpchr %in% paste0("chr",1:22) & snptype == "SNP")
clumped$code <- clumped %$% paste(snp, cpg)

mqtlcount <- group_by(clumped, cpg) %>% summarise(mqtls=n())
tab2l <- merge(tab2l, mqtlcount, by="cpg", all.x=TRUE)
tab2l$mqtls[is.na(tab2l$mqtls)] <- 0

p1 <- ggplot(tab2l, aes(x=as.factor(mqtls), y=value)) +
geom_boxplot(outlier.size=0.01, aes(fill=key)) +
theme(axis.text.x=element_text(angle=90)) +
labs(x="Number of mQTL per DNA methylation site", y="Heritability estimate", fill="Heritability\nstudy design")
ggsave(p1, file="images/mqtl_vs_h2.png", width=9, height=4)
ggsave(p1, file="images/mqtl_vs_h2.pdf", width=9, height=4)


tab2l %>% filter(key == "Twin") %>% lm(mqtls ~ value, data=.) %>% summary %>% coefficients

tab2l %>% group_by(key) %>% summarise(cor=cor(mqtls, value))

p1 <- ggplot(tab2l, aes(x=as.factor(round(r2*4, 1)/4), y=value)) +
geom_boxplot(outlier.size=0.01, aes(fill=key)) +
theme(axis.text.x=element_text(angle=90)) +
labs(x="Variance explained in GS replication", y="Heritability estimate", fill="Heritability\nstudy design")
ggsave(p1, file="images/r2_vs_h2.png", width=9, height=4)
ggsave(p1, file="images/r2_vs_h2.pdf", width=9, height=4)

write.csv(tab1, file="results/replication.csv")
write.csv(tab2, file="results/variance_explained.csv")

```



